File: app/src/main/java/com/dhsdevelopments/potato/clientapi/ImageUriRequestBody.java
package com.dhsdevelopments.potato.clientapi;

import android.content.Context;
import android.net.Uri;
import com.squareup.okhttp.MediaType;
import com.squareup.okhttp.RequestBody;
import okio.BufferedSink;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

public class ImageUriRequestBody extends RequestBody
{
    private Context context;
    private Uri imageUri;
    private MediaType mediaType;

    public ImageUriRequestBody( Context context, Uri imageUri ) {
        this.context = context;
        this.imageUri = imageUri;
        this.mediaType = MediaType.parse( context.getContentResolver().getType( imageUri ) );
    }

    @Override
    public MediaType contentType() {
        return mediaType;
    }

    @Override
    public void writeTo( BufferedSink sink ) throws IOException {
        try( InputStream in = context.getContentResolver().openInputStream( imageUri ) ) {
            OutputStream out = sink.outputStream();
            byte[] buf = new byte[16 * 1024];
            int n;
            while( (n = in.read( buf )) != -1 ) {
                out.write( buf, 0, n );
            }
        }
    }
}

File: app/src/main/java/com/dhsdevelopments/potato/clientapi/ImageUriRequestBody.kt
package com.dhsdevelopments.potato.clientapi

import android.content.Context
import android.net.Uri
import com.squareup.okhttp.MediaType
import com.squareup.okhttp.RequestBody
import okio.BufferedSink
import java.io.IOException

class ImageUriRequestBody(private val context: Context, private val imageUri: Uri) : RequestBody() {
    private val mediaType: MediaType

    init {
        this.mediaType = MediaType.parse(context.contentResolver.getType(imageUri))
    }

    override fun contentType(): MediaType {
        return mediaType
    }

    @Throws(IOException::class)
    override fun writeTo(sink: BufferedSink) {
        context.contentResolver.openInputStream(imageUri).use { inStream ->
            val out = sink.outputStream()
            val buf = ByteArray(16 * 1024)
            while(true) {
                val n = inStream.read(buf)
                if(n == -1) {
                    break
                }
                out.write(buf, 0, n)
            }
        }
    }
}
